web/middleware/request_logger.go

```diff
@@ -0,0 +1,101 @@
+package middleware
+
+import (
+	"bufio"
+	"context"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/inconshreveable/log15"
+	"github.com/pkg/errors"
+)
+
+var requestLoggerKey = &contextKey{"Logger"}
+
+func RequestLogger(logger log15.Logger) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		fn := func(w http.ResponseWriter, r *http.Request) {
+			// Add request id to logger if it is here (it should)
+			reqID := GetReqID(r)
+			l := logger
+			if reqID != "" {
+				l = logger.New("reqID", reqID)
+			}
+
+			// TODO: wrapResponseWriter to rereadit
+			wr := &wrappedResponse{0, 0, w}
+			t1 := time.Now()
+			defer func() {
+				t2 := time.Now()
+				t2.Sub(t1)
+
+				l.Info(fmt.Sprintf("//%s%s %s", r.Host, r.RequestURI, r.Proto),
+					"method", r.Method,
+					"path", r.URL.String(),
+					"duration", t2.Sub(t1),
+					"size", wr.bytes,
+					"status", wr.status)
+				// write the log
+			}()
+
+			r = r.WithContext(context.WithValue(r.Context(), requestLoggerKey, l))
+			next.ServeHTTP(wr, r)
+		}
+
+		return http.HandlerFunc(fn)
+	}
+}
+
+func GetLogger(r *http.Request) log15.Logger {
+	logger, _ := r.Context().Value(requestLoggerKey).(log15.Logger)
+	return logger
+}
+
+type wrappedResponse struct {
+	status int
+	bytes  int
+	http.ResponseWriter
+}
+
+var _ http.ResponseWriter = &wrappedResponse{}
```

Comment:
i'll remove a lot from middleware this is a little trick that should be put in tests (which is my bad) 

it tests that `wrappedResponse` implements the interface `http.ResponseWriter`